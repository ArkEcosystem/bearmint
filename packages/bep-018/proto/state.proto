syntax = "proto3";

import "google/protobuf/struct.proto";

option optimize_for = SPEED;

message BEP21MilestoneParameters {
  .google.protobuf.Struct application = 1;
  .google.protobuf.Struct consensus = 2;
  .google.protobuf.Struct denominations = 3;
  .google.protobuf.Struct mandatory_modules = 4;
  .google.protobuf.Struct transactions = 5;
}

message BEP21Milestone {
  uint64 block_number = 1;
  BEP21MilestoneParameters parameters = 2;
}
