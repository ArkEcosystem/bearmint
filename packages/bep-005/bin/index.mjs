#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
