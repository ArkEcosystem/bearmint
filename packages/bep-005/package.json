{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
