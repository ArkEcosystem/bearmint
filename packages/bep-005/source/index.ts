export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
