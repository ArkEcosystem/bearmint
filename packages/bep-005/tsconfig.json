{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
