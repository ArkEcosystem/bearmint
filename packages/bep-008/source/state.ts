import type { TxTestContext } from '@bearmint/bep-006'
import { milestone } from '@bearmint/bep-006'
import { canonicalizeJson, Paths, snakeCase } from '@bearmint/bep-009'
import type { KVStore } from '@bearmint/bep-013'
import { ContainerType } from '@bearmint/bep-013'
import { BEP21Milestone } from '@bearmint/bep-018'
import { makeMultiStore, makeState, TK_MILESTONE } from '@bearmint/bep-020'
import { makeSecureKVStore } from '@bearmint/bep-023'

import { fakeAccountRepositoryFactory } from './account.js'

interface StateStoreKVStores {
	account: KVStore
	accountMoniker: KVStore
	accountPublicKey: KVStore
	accountValidatorAddress: KVStore
	milestone: KVStore
	tx: KVStore
	txReceipt: KVStore
	world: KVStore
}

export async function fakeStateStore(context: TxTestContext) {
	context.container.bindValue(ContainerType.CommittedState, {
		getCommittedBlockAppHash() {
			return Buffer.from('deadbeef')
		},
		getMilestone() {
			return { ...milestone }
		},
	})

	context.container.bindValue(ContainerType.DeliverTxState, {
		getCommittedBlockAppHashpHash() {
			return Buffer.from('deadbeef')
		},
		getMilestone() {
			return { ...milestone }
		},
	})

	const result = await makeState(
		fakeAccountRepositoryFactory(context),
		await fakeMultiStore({
			milestone: await fakeMilestoneKVStore(),
		}),
	)

	await result.setMilestone(milestone as any)

	return result
}

export async function fakeMilestoneKVStore() {
	const result = await makeSecureKVStore().make()
	await result.set(
		TK_MILESTONE,
		BEP21Milestone.fromJson(
			canonicalizeJson({
				blockNumber: '1',
				parameters: {
					application: milestone.parameters.application,
					consensus: milestone.parameters.consensus,
					denominations: milestone.parameters.denominations,
					mandatoryModules: milestone.parameters.modules.mandatory,
					transactions: milestone.parameters.transactions,
				},
			}),
		).toBinary(),
	)

	return result
}

export async function fakeKVStores(tries?: Partial<StateStoreKVStores>) {
	return {
		account: tries?.account ?? (await makeSecureKVStore().make()),
		accountMoniker: tries?.accountMoniker ?? (await makeSecureKVStore().make()),
		accountPublicKey: tries?.accountPublicKey ?? (await makeSecureKVStore().make()),
		accountValidatorAddress: tries?.accountValidatorAddress ?? (await makeSecureKVStore().make()),
		milestone: tries?.milestone ?? (await makeSecureKVStore().make()),
		tx: tries?.tx ?? (await makeSecureKVStore().make()),
		txReceipt: tries?.txReceipt ?? (await makeSecureKVStore().make()),
		world: tries?.world ?? (await makeSecureKVStore().make()),
	}
}

export async function fakeMultiStore(tries?: Partial<StateStoreKVStores>) {
	const result = makeMultiStore({
		KVStoreFactory: makeSecureKVStore(),
		// @ts-ignore
		Logger: {
			debug() {
				//
			},
		},
		// @ts-ignore
		Paths: {
			data: `${Paths.BM_PATH_HOME}/data`,
		},
	})

	for (const [key, trie] of Object.entries(await fakeKVStores(tries))) {
		result.set(`@bearmint/bep-020/${snakeCase(key)}`, trie)
	}

	return result
}
