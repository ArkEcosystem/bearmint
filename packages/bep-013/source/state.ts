import type { abci, BEP88Validator } from '@bearmint/bep-018'

import type { Account, AccountRepository, AccountWithValidator } from './account.js'
import type { Milestone } from './milestone.js'
import type { KVStore, KVStoreOpts } from './store.js'

export interface StateStore {
	// Mutation
	copy(): Promise<StateStore>

	// KVStore Checkpoints
	hasCheckpoints(): boolean
	checkpoint(): Promise<void>
	commit(candidateState?: StateStore | undefined): Promise<{ height: bigint; root: Uint8Array }>
	revert(): Promise<void>

	// Stores
	getMultiStore(): MultiStore
	getTxReceiptStore(): KVStore
	getTxStore(): KVStore

	// Accounts
	getAccountRepository(): AccountRepository
	setAccountRepository(values: AccountRepository): void

	// Gas Consumption
	setConsumedGas(gas: bigint): void
	getConsumedGas(): bigint

	// Milestones
	getMilestone<T = object>(): Milestone<T>
	setMilestone(milestone: Milestone): Promise<void>

	// The root of the
	getAppHash(): Uint8Array

	// A candidate block is a block that has not yet reached consensus
	hasCandidateBlock(): boolean
	getCandidateBlock(): abci.RequestFinalizeBlock
	setCandidateBlock(proposedBlock: abci.RequestFinalizeBlock): void
	getCandidateBlockHash(): Uint8Array
	getCandidateBlockNumber(): bigint
	getCandidateBlockProposer(): Uint8Array

	// A committed block is a block that has reached consensus and has been persisted
	getCommittedBlockNumber(): Promise<bigint>
	getCommittedBlock(): Promise<abci.RequestFinalizeBlock>
	setCommittedBlock(block: abci.RequestFinalizeBlock): Promise<void>
	getCommittedBlockAppHash(): Promise<Uint8Array>

	// The last set of validator updates
	getValidatorUpdates(): Promise<abci.ValidatorUpdate[]>
	setValidatorUpdates(updates: abci.ValidatorUpdate[]): Promise<void>
}

export interface AccountRepositoryFactory {
	make(tries: {
		account: KVStore
		accountMoniker: KVStore
		accountPublicKey: KVStore
		accountValidatorAddress: KVStore
	}): AccountRepository
}

export interface MultiStore {
	add(module: string, opts?: KVStoreOpts): Promise<void>
	all(): Record<string, KVStore>
	checkpoint(): void
	copy(): MultiStore
	get(module: string): KVStore
	hasCheckpoints(): boolean
	revert(): Promise<void>
	roots(): Record<string, Buffer>
	set(module: string, trie: KVStore): void
}

export interface MultiStoreFactory {
	make(): MultiStore
}

export interface StakingService {
	allByValidator(validator: AccountWithValidator): Promise<BEP88Validator>
	keysByValidator(validator: AccountWithValidator): Promise<string[]>
	valuesByValidator(validator: AccountWithValidator): Promise<bigint[]>

	increaseStake(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	decreaseStake(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	multiplyStake(
		validator: AccountWithValidator,
		delegator: Account,
		percentage: number,
	): Promise<void>

	attach(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	detach(validator: AccountWithValidator, delegator: Account): Promise<void>
}

export interface StakingServiceFactory {
	make(state: StateStore): StakingService
}
