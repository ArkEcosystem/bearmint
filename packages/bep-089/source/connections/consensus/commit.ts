import { bytesToHex, canonicalizeJson } from '@bearmint/bep-009'
import type { ABCIController, Cradle } from '@bearmint/bep-013'
import type { abci } from '@bearmint/bep-018'
import { getModuleMilestone } from '@bearmint/bep-021'

import { resetState } from './utils.js'

/**
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/apps.md#consensus-connection
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/abci.md#commit
 *
 * @see https://cosmos-network.gitbooks.io/cosmos-academy/content/cosmos-for-developers/tendermint/abci-protocol.html
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/abci.md#info
 */
export function makeCommit(
	cradle: Cradle,
): ABCIController<abci.RequestCommit, abci.ResponseCommit> {
	return {
		async execute(_request) {
			const { height, root } = await cradle.CommittedState.commit(cradle.ExecuteTxState)

			// We reset here because `CheckTx` will be called next if there are pending txs.
			await resetState(cradle)

			if (cradle.Logger !== undefined) {
				/* eslint-disable sort-keys-fix/sort-keys-fix */
				const msg = {
					module: '@bearmint/bep-089',
					action: 'commit_roots',
					blockNumber: (await cradle.CommittedState.getCommittedBlockNumber()).toString(),
				}

				for (const [submodule, value] of Object.entries(
					cradle.CommittedState.getMultiStore().roots(),
				)) {
					cradle.Logger.debug('%s', { ...msg, submodule, root: bytesToHex(value) })
				}

				cradle.Logger.debug('%s', {
					module: '@bearmint/bep-089',
					action: 'commit_app_hash',
					root: bytesToHex(root),
				})
				/* eslint-enable sort-keys-fix/sort-keys-fix */
			}

			// Dump the genesis and milestone parameters
			cradle.DataSink.put(
				'@bearmint/bep-097',
				{
					id: 'genesis',
					type: 'entity',
				},
				{
					key: 'genesis',
					type: 'configuration',
					value: canonicalizeJson(cradle.GenesisParameters),
				},
			)

			cradle.DataSink.put(
				'@bearmint/bep-021',
				{
					id: 'milestones',
					type: 'entity',
				},
				{
					key: 'milestones',
					type: 'configuration',
					value: canonicalizeJson(cradle.Milestones),
				},
			)

			/**
			 * Use RetainHeight with caution!
			 * -----------------------------------------------------------------
			 * If all nodes in the network remove historical blocks then this
			 * data is permanently lost, and no new nodes will be able to join
			 * the network and bootstrap. Historical blocks may also be required
			 * for other purposes, e.g. auditing, replay of non-persisted
			 * heights, light client verification, and so on.
			 */
			let retainHeight = BigInt(0)

			const { retainBlocks } = getModuleMilestone<{ retainBlocks: number }>(
				cradle.ServiceProviderRepository,
				cradle.CommittedState,
				'@bearmint/bep-089',
			)

			if (retainBlocks > 0) {
				retainHeight = BigInt(height - BigInt(retainBlocks + 1))
			}

			return {
				data: root,
				retainHeight,
			}
		},
	}
}
