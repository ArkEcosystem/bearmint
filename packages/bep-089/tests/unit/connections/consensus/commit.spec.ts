import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import { makeContainer } from '@bearmint/bep-014'
import { tmpdir } from 'node:os'

import { fakeStateStore } from '../../../state.js'
import { makeCommit } from '../../../../source/connections/consensus/commit.js'
import { abci } from '@bearmint/bep-018'
import { config } from '@bearmint/bep-006'

describe('Commit', () => {
	it('it should respond without any errors', async () => {
		const CommittedState = await fakeStateStore()
		CommittedState.setCandidateBlock(new abci.RequestBeginBlock({ header: { height: BigInt(1) } }))
		await CommittedState.setCommittedBlock(
			new abci.RequestBeginBlock({ header: { height: BigInt(1) } }),
		)
		CommittedState.setMilestone(config.milestones[0])

		await CommittedState.checkpoint()
		await CommittedState.checkpoint()

		expect(
			await makeCommit({
				CheckTxState: await CommittedState.copy(),
				Container: makeContainer(),
				DataSink: { put() {} },
				DeliverTxState: await CommittedState.copy(),
				CommittedState,
				GenesisParameters: {},
				Milestones: config.milestones,
				Paths: { data: tmpdir() },
				ServiceProviderRepository: {
					get() {
						return {
							version() {
								return '0.0.0'
							},
						}
					},
				},
			}).execute({}),
		).toMatchSnapshot()
	})
})
