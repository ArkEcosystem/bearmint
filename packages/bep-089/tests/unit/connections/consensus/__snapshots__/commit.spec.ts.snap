// Vitest Snapshot v1

exports[`Commit > it should respond without any errors 1`] = `
{
  "data": {
    "data": [
      179,
      143,
      54,
      217,
      57,
      59,
      6,
      240,
      181,
      167,
      128,
      184,
      71,
      120,
      65,
      244,
      243,
      92,
      31,
      191,
      78,
      252,
      104,
      198,
      154,
      52,
      43,
      148,
      247,
      230,
      5,
      17,
    ],
    "type": "Buffer",
  },
  "retainHeight": 0n,
}
`;
